ocean_lib/ocean/ocean_assets.py

```diff
@@ -322,48 +375,88 @@ def create(
-            )
-            document = encrypt_response.text
+        return asset
 
-        # Encrypted & compressed
-        else:
-            flags = bytes([3])
-            # Compress DDO
-            compressed_document = lzma.compress(ddo_bytes)
+    def update(
+        self,
+        asset: Asset,
+        publisher_wallet: Wallet,
+        provider_uri: Optional[str] = None,
+        encrypt_flag: Optional[bool] = False,
+        compress_flag: Optional[bool] = False,
+    ) -> str:
+        """Update an asset on-chain.
 
-            # Encrypt DDO
-            encrypt_response = DataServiceProvider.encrypt(
-                objects_to_encrypt=compressed_document,
-                encrypt_endpoint=f"{provider_uri}/api/services/encrypt",
+        :param asset: The updated asset to update on-chain
+        :param publisher_wallet: Wallet of the publisher updating this asset.
+        :param provider_uri: str URL of service provider. This will be used as base to construct the serviceEndpoint for the `access` (download) service
+        :param encrypt_flag: bool for encryption of the DDO.
+        :param compress_flag: bool for compression of the DDO.
+        :return: update tx_result
+        """
+
+        # Metadata sanity check
+        assert isinstance(
+            asset.metadata, dict
+        ), f"Expected asset.metadata of type dict, got {type(asset.metadata)}"
+
+        asset_type = asset.metadata.get("type")
+
+        assert asset_type in (
+            "dataset",
+            "algorithm",
+        ), f"Invalid/unsupported asset type {asset_type}"
+
+        assert "name" in asset.metadata, "Must have name in metadata."
+
+        if not provider_uri:
+            provider_uri = DataServiceProvider.get_url(self._config)
+
+        address = get_address_of_type(self._config, ERC721FactoryContract.CONTRACT_NAME)
+        erc721_factory = ERC721FactoryContract(self._web3, address)
+        erc721_address = asset.nft_address
+
+        # Verify nft address
+        if not erc721_factory.verify_nft(erc721_address):
+            raise ContractNotFound(
+                f"NFT address {erc721_address} is not found in the ERC721Factory events."
             )
 
-            document = encrypt_response.text
+        assert erc721_address, "nft_address is required for publishing a dataset asset."
+        erc721_token = ERC721Token(self._web3, erc721_address)
 
-        _ = erc721_token.set_metadata(
+        assert asset.chain_id == self._web3.eth.chain_id, "Chain id mismatch."
+
+        # Validation by Aquarius
+        validation_result, validation_errors = self.validate(asset)
+        if not validation_result:
+            msg = f"Asset has validation errors: {validation_errors}"
+            logger.error(msg)
+            raise ValueError(msg)
+
+        document, flags, ddo_hash = self.process_ddo(
+            asset, provider_uri, encrypt_flag, compress_flag
+        )
+
+        return erc721_token.set_metadata(
```

Comment:
I think we should return the asset (and wait for it) instead. What was the return type for v3?